sentry-tower/src/http.rs

```diff
@@ -0,0 +1,146 @@
+///
+/// The Service can also optionally start a new performance monitoring transaction
+/// based on incoming distributed tracing headers.
+#[derive(Clone)]
+pub struct SentryHttpService<S> {
+    service: S,
+    start_transaction: bool,
+}
+
+impl<S> Layer<S> for SentryHttpLayer {
+    type Service = SentryHttpService<S>;
+
+    fn layer(&self, service: S) -> Self::Service {
+        Self::Service {
+            service,
+            start_transaction: self.start_transaction,
+        }
+    }
+}
+
+/// The Future returned from [`SentryHttpService`]
+#[pin_project::pin_project]
+pub struct SentryHttpFuture<F> {
+    transaction: Option<(
+        sentry_core::TransactionOrSpan,
+        Option<sentry_core::TransactionOrSpan>,
+    )>,
+    #[pin]
+    future: F,
+}
+
+impl<F> Future for SentryHttpFuture<F>
+where
+    F: Future,
+{
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let slf = self.project();
+        match slf.future.poll(cx) {
+            Poll::Ready(res) => {
+                if let Some((transaction, parent_span)) = slf.transaction.take() {
+                    transaction.finish();
+                    sentry_core::configure_scope(|scope| scope.set_span(parent_span));
+                }
+                Poll::Ready(res)
+            }
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+impl<S, Body> Service<Request<Body>> for SentryHttpService<S>
+where
+    S: Service<Request<Body>>,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = SentryHttpFuture<S::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.service.poll_ready(cx)
+    }
+
+    fn call(&mut self, request: Request<Body>) -> Self::Future {
+        let transaction = sentry_core::configure_scope(|scope| {
+            let sentry_req = sentry_core::protocol::Request {
+                method: Some(request.method().to_string()),
+                url: request.uri().to_string().parse().ok(),
+                headers: request
+                    .headers()
+                    .into_iter()
+                    .map(|(header, value)| {
+                        (
+                            header.to_string(),
+                            value.to_str().unwrap_or_default().into(),
+                        )
+                    })
+                    .collect(),
+                ..Default::default()
+            };
+            scope.add_event_processor(move |mut event| {
+                if event.request.is_none() {
+                    event.request = Some(sentry_req.clone());
+                }
+                Some(event)
+            });
+
+            if self.start_transaction {
+                let headers = request.headers().into_iter().flat_map(|(header, value)| {
+                    value.to_str().ok().map(|value| (header.as_str(), value))
+                });
```

Comment:
similar here, i'm fairly certain you don't need to have all headers and can filter on only the sentry-tracing ones

Reply:
thats what `TransactionContext::continue_from_headers` does internally, the API is part of the unified spec:
https://develop.sentry.dev/sdk/performance/#new-span-and-transaction-classes
> TransactionContext should have a static/ctor method called continueFromHeaders(headerMap) which is really just a thin wrapper around fromSentryTrace(headerMap.get("sentry-trace")) right now. This should be preferred by integration/framework-sdk authors over fromSentryTrace as it hides the exact header names used deeper in the core sdk, and leaves opportunity for using additional headers (from the W3C) in the future without changing all integrations.

Reply:
this is true, but we know which headers `continue_from_headers` actually uses, so we can filter them up-front can we not?  though maybe if this stays as an iterator it doesn't have that much of an impact?